test/runtime/test_apply_requirement_modifications.rb

```diff
@@ -304,6 +304,202 @@ module Runtime
                 assert other_requirement_task.failed?
             end
 
+            it "resolves all requirements that were added to the plan in successive " \
+               "#apply_requirement_modifications calls" do
+                task_m = Composition.new_submodel
+                task_m2 = Composition.new_submodel
+                task_m3 = Composition.new_submodel
+                reqs = [task_m, task_m2, task_m3].map do |t|
+                    requirement_task =
+                        plan.add_permanent_task(t.to_instance_requirements.as_plan)
+                    requirement_task.planning_task
+                end
+                execute { reqs.first.start! }
+                execute { Runtime.apply_requirement_modifications(plan) }
+                execute { reqs[1].start! }
+                execute { Runtime.apply_requirement_modifications(plan) }
+                execute { reqs[2].start! }
+                execute { Runtime.apply_requirement_modifications(plan) }
+                assert plan.syskit_has_async_resolution?
+
+                execute { plan.syskit_join_current_resolution }
+                reqs.each do |req|
+                    assert req.resolution_success?
+                end
+            end
+
+            it "while using force, cancels pending resolutions and executes the last " \
+               "one" do
+                task_m = Composition.new_submodel
+                requirement_task =
+                    plan.add_permanent_task(task_m.to_instance_requirements.as_plan)
+                requirement_task = requirement_task.planning_task
+                execute { requirement_task.start! }
+                execute { Runtime.apply_requirement_modifications(plan, force: true) }
+                assert plan.syskit_has_async_resolution?
+                FlexMock.use(plan) do |mock|
+                    mock.should_receive(syskit_finished_async_resolution?: false)
+                    execute { Runtime.apply_requirement_modifications(plan, force: true) }
+                end
+                assert plan.syskit_current_resolution.cancelled?
+                execute { Runtime.apply_requirement_modifications(plan, force: true) }
```

Comment:
You need to join cancelled resolutions to create new heisenbugs. I suggest doing a complete read with only this in mind ...